tool_path_planner/src/halfedge_boundary_finder.cpp

```diff
@@ -0,0 +1,327 @@
+
+      return n +  v;
+    });
+    p.getNormalVector3fMap() = average_normal.normalized();
+  }
+}
+
+namespace tool_path_planner
+{
+HalfEdgeBoundaryFinder::HalfEdgeBoundaryFinder()
+{
+
+}
+
+HalfEdgeBoundaryFinder::~HalfEdgeBoundaryFinder()
+{
+
+}
+
+void HalfEdgeBoundaryFinder::setInput(pcl::PolygonMesh::ConstPtr mesh)
+{
+  mesh_ = mesh;
+}
+
+void HalfEdgeBoundaryFinder::setInput(const shape_msgs::Mesh& mesh)
+{
+  pcl::PolygonMesh::Ptr pcl_mesh = boost::make_shared<pcl::PolygonMesh>();
+  noether_conversions::convertToPCLMesh(mesh,*pcl_mesh);
+  setInput(pcl_mesh);
+}
+
+boost::optional<std::vector<geometry_msgs::PoseArray> >
+HalfEdgeBoundaryFinder::generate(const tool_path_planner::HalfEdgeBoundaryFinder::Config& config)
+{
+  using namespace pcl;
+  CONSOLE_BRIDGE_logInform("Input mesh has %lu polygons",mesh_->polygons.size());
+  boost::optional<TraingleMesh> mesh = createTriangleMesh(*mesh_);
+  if(!mesh)
+  {
+    CONSOLE_BRIDGE_logError("Failed to create triangle mesh");
+    return boost::none;
+  }
+
+  if(mesh->getVertexDataCloud().empty())
+  {
+    CONSOLE_BRIDGE_logError("Generated Triangle mesh contains no data");
+    return boost::none;
+  }
+
+  std::vector<TraingleMesh::HalfEdgeIndices> boundary_he_indices;
+  getBoundBoundaryHalfEdges(mesh.get(), boundary_he_indices);
+
+  // initializing octree
+  PointCloud<PointNormal>::Ptr input_cloud = boost::make_shared<PointCloud<PointNormal>>();
+  PointCloud<PointXYZ>::Ptr input_points = boost::make_shared<PointCloud<PointXYZ>>();
+  noether_conversions::convertToPointNormals(*mesh_, *input_cloud);
+  pcl::copyPointCloud(*input_cloud, *input_points );
+
+
+  pcl::KdTreeFLANN<pcl::PointXYZ> kdtree;
+  kdtree.setInputCloud (input_points);
+
+  // traversing half edges list
+  std::vector<geometry_msgs::PoseArray> boundary_poses;
+
+
+  for (std::vector<TraingleMesh::HalfEdgeIndices>::const_iterator boundary = boundary_he_indices.begin(),
+                                                                  b_end = boundary_he_indices.end();
+       boundary != b_end; ++boundary)
+  {
+    geometry_msgs::PoseArray bound_segment_poses;
+    PointCloud<PointNormal> bound_segment_points;
+    for (TraingleMesh::HalfEdgeIndices::const_iterator edge = boundary->begin(), edge_end = boundary->end();
+         edge != edge_end; ++edge)
+    {
+      TraingleMesh::VertexIndex vertex = mesh->getOriginatingVertexIndex(*edge);
+      if(mesh->getVertexDataCloud().size() <= vertex.get())
+      {
+        CONSOLE_BRIDGE_logError("Vertex index exceeds size of vertex list");
+        return boost::none;
+      }
+      std::size_t source_idx = mesh->getVertexDataCloud()[vertex.get()];
+
+      if(input_cloud->size() <= source_idx)
+      {
+        CONSOLE_BRIDGE_logError("Point index exceeds size of mesh point cloud");
+        return boost::none;
+      }
+
+      bound_segment_points.push_back((*input_cloud)[source_idx]);
+    }
+
+    // decimating
+    if(config.min_point_dist >  1e-6 )
```

Comment:
should this be a parameter?

Reply:
I'll make it a global const but it does not need to be a parameter